File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int res = 0;
    int maxPathSum(TreeNode* root) {
        res = root->val;
        dfs(root);
        return res;
    }
    int dfs(TreeNode* root) {
        if (root==nullptr) return 0;
        int leftv = dfs(root->left);
        int rightv = dfs(root->right);
        int leftmax = max(leftv,0);
        int rightmax = max(rightv,0);
        int temp = leftmax + rightmax;
        res = max(res,root->val + temp);
        return root->val + max(leftmax,rightmax);
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-3.py
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height:
            return 0
        l,r = 0,len(height)-1
        leftmax,rmax = height[l],height[r]
        res=0
        while l < r :
            if leftmax < rmax:
                l+=1
                leftmax = max(leftmax,height[l])
                res += leftmax - height[l]
            else:
                r-=1
                rmax = max(rmax,height[r])
                res+=rmax - height[r]
        return res